gnosis/safe/safe.py

```diff
@@ -207,423 +139,130 @@ def _deploy_master_contract(
+        :param gas_price: Gas Price
+        :param gas_token: Gas Token, to use token instead of ether for the gas
+        :return: `True` if enough funds, `False` otherwise
         """
+        if gas_token == NULL_ADDRESS:
+            balance = self.ethereum_client.get_balance(self.address)
+        else:
+            balance = self.ethereum_client.erc20.get_balance(self.address, gas_token)
+        return balance >= (safe_tx_gas + base_gas) * gas_price
 
-        return cls._deploy_master_contract(
-            ethereum_client, deployer_account, get_safe_V1_3_0_contract
-        )
-
-    @classmethod
-    def deploy_master_contract_v1_1_1(
-        cls, ethereum_client: EthereumClient, deployer_account: LocalAccount
-    ) -> EthereumTxSent:
+    def estimate_tx_base_gas(
+        self,
+        to: str,
+        value: int,
+        data: bytes,
+        operation: int,
+        gas_token: str,
+        estimated_tx_gas: int,
+    ) -> int:
         """
-        Deploy master contract v1.1.1. Takes deployer_account (if unlocked in the node) or the deployer private key
-        Safe with version > v1.1.1 doesn't need to be initialized as it already has a constructor
+        Calculate gas costs that are independent of the transaction execution(e.g. base transaction fee,
+        signature check, payment of the refund...)
 
-        :param ethereum_client:
-        :param deployer_account: Ethereum account
-        :return: deployed contract address
+        :param to:
+        :param value:
+        :param data:
+        :param operation:
+        :param gas_token:
+        :param estimated_tx_gas: gas calculated with `estimate_tx_gas`
+        :return:
         """
+        data = data or b""
+        safe_contract = self.contract
+        threshold = self.retrieve_threshold()
+        nonce = self.retrieve_nonce()
 
-        return cls._deploy_master_contract(
-            ethereum_client, deployer_account, get_safe_V1_1_1_contract
+        # Every byte == 0 -> 4  Gas
+        # Every byte != 0 -> 16 Gas (68 before Istanbul)
+        # numbers < 256 (0x00(31*2)..ff) are 192 -> 31 * 4 + 1 * GAS_CALL_DATA_BYTE
+        # numbers < 65535 (0x(30*2)..ffff) are 256 -> 30 * 4 + 2 * GAS_CALL_DATA_BYTE
+
+        # Calculate gas for signatures
+        # (array count (3 -> r, s, v) + ecrecover costs) * signature count
+        # ecrecover for ecdsa ~= 4K gas, we use 6K
+        ecrecover_gas = 6000
+        signature_gas = threshold * (
+            1 * GAS_CALL_DATA_BYTE + 2 * 32 * GAS_CALL_DATA_BYTE + ecrecover_gas
         )
 
-    @staticmethod
-    def deploy_master_contract_v1_0_0(
-        ethereum_client: EthereumClient, deployer_account: LocalAccount
-    ) -> EthereumTxSent:
-        """
-        Deploy master contract. Takes deployer_account (if unlocked in the node) or the deployer private key
+        safe_tx_gas = estimated_tx_gas
+        base_gas = 0
+        gas_price = 1
+        gas_token = gas_token or NULL_ADDRESS
```

Comment:
Can be the default NULL_ADDRESS value defined in the method header?

Reply:
This could break some behaviour, also `NULL_ADDRESS` is our current implementation for Solidity, while someone that's just using Python can say that the `gas_token` is `None`, and that should be fair in my opinion. We are just "translating"

Reply:
We can change that but I wouldn't do it in this PR